binderhub/repoproviders.py

```diff
@@ -448,6 +448,73 @@ def get_build_slug(self):
         return f"hydroshare-{self.record_id}"
 
 
+class CKANProvider(RepoProvider):
+    """Provide contents of a CKAN dataset
+    Users must provide a spec consisting of the CKAN dataset URL.
+    """
+
+    name = Unicode("CKAN")
+
+    display_name = "CKAN dataset"
+
+    url_regex = r"/dataset/[a-z0-9_\\-]*$"
+
+    labels = {
+        "text": "CKAN dataset URL (https://demo.ckan.org/dataset/sample-dataset-1)",
+        "tag_text": "Git ref (branch, tag, or commit)",
+        "ref_prop_disabled": True,
+        "label_prop_disabled": True,
+    }
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.repo = urllib.parse.unquote(self.spec)
+
+    async def get_resolved_ref(self):
+        parsed_repo = urlparse(self.repo)
+        self.dataset_id = parsed_repo.path.rsplit("/", maxsplit=1)[1]
+
+        client = AsyncHTTPClient()
+
+        api = parsed_repo._replace(
+            path=re.sub(self.url_regex, "/api/3/action/", parsed_repo.path)
```

Comment:
Same comment about using regexes as in https://github.com/jupyterhub/repo2docker/pull/1336#discussion_r1536545745.

Reply:
Thanks. I have made changes according to your suggestion.